syntax = "proto3";

package ocr;

message BoundingBox {
  float x1 = 1;
  float y1 = 2;
  float x2 = 3;
  float y2 = 4;
}

message Polygon {
  repeated float x = 1;
  repeated float y = 2;
}

message DetectedElement {
  string content = 1;
  BoundingBox bbox = 2;
  Polygon polygon = 3;
  float confidence = 4;
  int32 reading_order = 5;
}

message LayoutElement {
  enum ElementType {
    BLANK = 0;
    TEXT = 1;
    TEXT_INLINE_MATH = 2;
    CODE = 3;
    SECTION_HEADER = 4;
    CAPTION = 5;
    FOOTNOTE = 6;
    EQUATION = 7;
    LIST_ITEM = 8;
    PAGE_FOOTER = 9;
    PAGE_HEADER = 10;
    PICTURE = 11;
    FIGURE = 12;
    TABLE = 13;
    FORM = 14;
    TABLE_OF_CONTENTS = 15;
    HANDWRITING = 16;
  }
  ElementType type = 1;
  DetectedElement element = 2;
}

message TableCell {
  DetectedElement content = 1;
  int32 row_id = 2;
  int32 col_id = 3;
  int32 row_span = 4;
  int32 col_span = 5;
}

message Table {
  repeated TableCell cells = 1;
  int32 num_rows = 2;
  int32 num_cols = 3;
  BoundingBox bbox = 4;
}

message ImageRequest {
  bytes image_data = 1;
  string image_format = 2;
}

message DocumentAnalysisResponse {
  repeated DetectedElement text_lines = 1;
  repeated DetectedElement vertical_lines = 2;
  repeated LayoutElement layout_elements = 3;
  repeated Table tables = 4;
  repeated DetectedElement reading_order = 5;
}

service DocumentAnalysis {
  rpc AnalyzeDocument(ImageRequest) returns (DocumentAnalysisResponse) {}
}